Планирование – залог успеха

kind-chigik opened the pull request:
---
:mortar_board: [Планирование – залог успеха](https://up.htmlacademy.ru/yii/1/user/669277/tasks/5)

kind-chigik committed:
Подготовлена схема БД

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+USE taskforce;
+
+CREATE TABLE city (
+  id INT AUTO_INCREMENT PRIMARY KEY,
```
Для id имеет смысл всегда использовать `UNSIGNED` варианты типов (`INT UNSIGNED`, в данном случае)

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+CREATE TABLE city (
+  id INT AUTO_INCREMENT PRIMARY KEY,
+  title VARCHAR(64) NOT NULL,
+  coordinates INT NOT NULL
```
мм, широта и долгота должны храниться. Как ты их в одно число засунешь? )

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+  title VARCHAR(128) NOT NULL UNIQUE
+);
+
+CREATE TABLE role (
```
Роль, это заказчик или исполнитель? Если да - то в БД ее хранить не надо (как и статусы), поскольку динамически добавлять/удалять роли не получится, для этого нужно будет менять логику приложения. А в БД должны быть только динамические данные, которые можно свободно добавлять/удалять.

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+  title VARCHAR(64) NOT NULL UNIQUE
+);
+
+CREATE TABLE portfolio (
```
Тут хранятся фото работ, надо переименовать соответствующе

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+
+CREATE TABLE portfolio (
+  id INT AUTO_INCREMENT PRIMARY KEY,
+  image VARCHAR(255) NOT NULL
```
Можно переименовать в image_path, image_url или типа того. Будет более очевидно, что тут именно путь к картинке.

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+  rank INT DEFAULT '0',
+  city_id INT NOT NULL,
+  role_id INT NOT NULL,
+  status_id INT NOT NULL,
```
Стоит сделать поле статус строковым и можно использовать [ENUM](https://dev.mysql.com/doc/refman/5.7/en/enum.html) для ограничения возможных значений в status.

Ovsyanka commented on schema.sql:
```diff
@@ -0,0 +1,129 @@
+  FOREIGN KEY (city_id) REFERENCES city (id),
+  FOREIGN KEY (role_id) REFERENCES role (id),
+  FOREIGN KEY (status_id) REFERENCES status_user (id),
+  FOREIGN KEY (photo_work_id) REFERENCES portfolio (id)
```
Так получается, что к пользователю может быть привязана только одна фотография. А по ТЗ их может быть много.

Ovsyanka reviewed (changes requested):
Ну что ж - в целом очень даже хорошо. Косяк, пожалуй, один - некорректные связи между таблицами. Устно мы это проговорили. 

Другие замечания, которые я написал по конкретным полям относятся и ко всем аналогичным случаям - их ты найди сама.

По поводу ограничений внешнего ключа - можно указать, что делать при удалении связанных записей. По умолчанию, удаление записи, на которую что-то ссылается - запрещается. Но можно сделать так, чтобы это было разрешено. а связанные с ней записи тоже удалялись автоматически. Можно ничего не менять и оставить как есть, просто почитай об этом, чтобы знать что такое есть https://metanit.com/sql/mysql/2.5.php

kind-chigik committed:
Исправлены ошибки, внесены корректировки

Названия таблиц городов переименованы во множественное число.
В таблицу "cities" добавлены поля для хранения широты и долготы города.
Удалены таблицы для хранения статусов и ролей пользователей.
Установлены правильные связи между таблицами.
Убрано поле "another_contact", т.к. в нем пока нет необходимости.
Для полей, содержащих id добавлено значение UNSIGNED.
Добавлены действия, для зависимых таблиц в случае удаления связанных строк в главной таблице.
Скорректированы ограничения полей.

kind-chigik committed:
Исправлены недочеты

В таблицу "tasks" добавлено поле для хранения статуса задачи.
Добавлена промежуточная таблица "user_specializations", которая устанавливает связи между таблицами "users" и "specializations".

kind-chigik committed:
Исправлены значения поля "status", скорректированы действия для ON DELETE
